File: 43226/server/internal/repository/exhibition.go
package repository

import (
	"time"

	"gorm.io/gorm"

	"museum-server/internal/dto"
	"museum-server/internal/models"
)

type ExhibitionRepository struct {
	db *gorm.DB
}

func NewExhibitionRepository(db *gorm.DB) *ExhibitionRepository {
	return &ExhibitionRepository{db: db}
}

func (r *ExhibitionRepository) Create(exhibition *models.Exhibition) error {
	return r.db.Create(exhibition).Error
}

func (r *ExhibitionRepository) FindByID(id uint) (*models.Exhibition, error) {
	var exhibition models.Exhibition
	err := r.db.Preload("Collections").Preload("Museum").First(&exhibition, id).Error
	if err != nil {
		return nil, err
	}
	return &exhibition, nil
}

func (r *ExhibitionRepository) Update(exhibition *models.Exhibition) error {
	return r.db.Save(exhibition).Error
}

func (r *ExhibitionRepository) Delete(id uint) error {
	return r.db.Transaction(func(tx *gorm.DB) error {
		if err := tx.Where("exhibition_id = ?", id).Delete(&models.ExhibitionCollection{}).Error; err != nil {
			return err
		}
		return tx.Delete(&models.Exhibition{}, id).Error
	})
}

func (r *ExhibitionRepository) List(query *dto.ExhibitionListQuery) ([]models.Exhibition, int64, error) {
	var exhibitions []models.Exhibition
	var total int64

	db := r.db.Model(&models.Exhibition{})

	if query.Keyword != "" {
		db = db.Where("title LIKE ? OR description LIKE ?",
			"%"+query.Keyword+"%", "%"+query.Keyword+"%")
	}
	if query.Status != "" {
		db = db.Where("status = ?", query.Status)
	}
	if !query.StartDate.IsZero() {
		db = db.Where("start_date >= ?", query.StartDate)
	}
	if !query.EndDate.IsZero() {
		db = db.Where("end_date <= ?", query.EndDate)
	}

	db.Count(&total)
	db.Preload("Collections").Preload("Museum").
		Offset((query.Page - 1) * query.PageSize).Limit(query.PageSize).
		Order("created_at DESC").
		Find(&exhibitions)

	return exhibitions, total, nil
}

func (r *ExhibitionRepository) AddCollections(exhibitionID uint, collectionIDs []uint, sortOrders map[uint]int) error {
	return r.db.Transaction(func(tx *gorm.DB) error {
		for _, cid := range collectionIDs {
			sortOrder := 0
			if so, ok := sortOrders[cid]; ok {
				sortOrder = so
			}
			ec := models.ExhibitionCollection{
				ExhibitionID: exhibitionID,
				CollectionID: cid,
				SortOrder:    sortOrder,
				CreatedAt:    time.Now(),
			}
			if err := tx.Create(&ec).Error; err != nil {
				return err
			}
		}
		return nil
	})
}

func (r *ExhibitionRepository) RemoveCollections(exhibitionID uint, collectionIDs []uint) error {
	return r.db.Where("exhibition_id = ? AND collection_id IN ?", exhibitionID, collectionIDs).
		Delete(&models.ExhibitionCollection{}).Error
}

func (r *ExhibitionRepository) GetCollections(exhibitionID uint) ([]models.Collection, error) {
	var collections []models.Collection
	err := r.db.Joins("JOIN exhibition_collections ON collections.id = exhibition_collections.collection_id").
		Where("exhibition_collections.exhibition_id = ?", exhibitionID).
		Order("exhibition_collections.sort_order ASC").
		Find(&collections).Error
	return collections, err
}

func (r *ExhibitionRepository) CreateTimeSlot(slot *models.TimeSlot) error {
	return r.db.Create(slot).Error
}

func (r *ExhibitionRepository) FindTimeSlotByID(id uint) (*models.TimeSlot, error) {
	var slot models.TimeSlot
	err := r.db.First(&slot, id).Error
	if err != nil {
		return nil, err
	}
	return &slot, nil
}

func (r *ExhibitionRepository) ListTimeSlots(exhibitionID uint, date time.Time) ([]models.TimeSlot, error) {
	var slots []models.TimeSlot
	db := r.db.Where("exhibition_id = ?", exhibitionID)
	if !date.IsZero() {
		db = db.Where("date = ?", date)
	}
	err := db.Order("start_time ASC").Find(&slots).Error
	return slots, err
}

func (r *ExhibitionRepository) UpdateTimeSlot(slot *models.TimeSlot) error {
	return r.db.Save(slot).Error
}

func (r *ExhibitionRepository) IncrementViewCount(id uint) error {
	return r.db.Model(&models.Exhibition{}).Where("id = ?", id).
		UpdateColumn("view_count", gorm.Expr("view_count + 1")).Error
}

File: 43226/web/src/api/exhibition.ts
import request from '@/utils/request'
import type {
  Exhibition,
  Collection,
  TimeSlot,
  ApiResponse,
  PageResult
} from '@/types'

export interface ExhibitionQuery {
  page?: number
  page_size?: number
  keyword?: string
  status?: string
  start_date?: string
  end_date?: string
}

export const listExhibitions = (params?: ExhibitionQuery) => {
  return request.get<any, ApiResponse<PageResult<Exhibition>>>('/exhibitions', { params })
}

export const getHotExhibitions = () => {
  return request.get<any, ApiResponse<Exhibition[]>>('/exhibitions/hot')
}

export const getExhibition = (id: number) => {
  return request.get<any, ApiResponse<Exhibition>>(`/exhibitions/${id}`)
}

export const createExhibition = (data: Partial<Exhibition> & { collection_ids?: number[] }) => {
  return request.post<any, ApiResponse<Exhibition>>('/exhibitions', data)
}

export const updateExhibition = (id: number, data: Partial<Exhibition> & { collection_ids?: number[] }) => {
  return request.put<any, ApiResponse<void>>(`/exhibitions/${id}`, data)
}

export const deleteExhibition = (id: number) => {
  return request.delete<any, ApiResponse<void>>(`/exhibitions/${id}`)
}

export const addExhibitionCollections = (id: number, collection_ids: number[]) => {
  return request.post<any, ApiResponse<void>>(`/exhibitions/${id}/collections`, { collection_ids })
}

export const removeExhibitionCollections = (id: number, collection_ids: number[]) => {
  return request.delete<any, ApiResponse<void>>(`/exhibitions/${id}/collections`, { data: { collection_ids } })
}

export const getExhibitionCollections = (id: number) => {
  return request.get<any, ApiResponse<Collection[]>>(`/exhibitions/${id}/collections`)
}

export const createTimeSlot = (exhibitionId: number, data: Partial<TimeSlot>) => {
  return request.post<any, ApiResponse<TimeSlot>>(`/exhibitions/${exhibitionId}/time-slots`, data)
}

export const batchCreateTimeSlots = (data: {
  exhibition_id: number
  start_date: string
  end_date: string
  start_time: string
  end_time: string
  interval: number
  max_capacity: number
}) => {
  return request.post<any, ApiResponse<void>>('/exhibitions/time-slots/batch', data)
}

export const listTimeSlots = (exhibitionId: number, date?: string) => {
  return request.get<any, ApiResponse<TimeSlot[]>>(`/exhibitions/${exhibitionId}/time-slots`, {
    params: { date }
  })
}
